File: Medium/Add two numbers represented by linked lists/add-two-numbers-represented-by-linked-lists.cpp
//{ Driver Code Starts
// driver

#include <bits/stdc++.h>
using namespace std;

/* Linked list Node */
struct Node {
    int data;
    struct Node* next;
    Node(int x) {
        data = x;
        next = NULL;
    }
};

struct Node* buildList(int size)
{
    int val;
    cin>> val;
    
    Node* head = new Node(val);
    Node* tail = head;
    
    for(int i=0; i<size-1; i++)
    {
        cin>> val;
        tail->next = new Node(val);
        tail = tail->next;
    }
    
    return head;
}

void printList(Node* n)
{
    while(n)
    {
        cout<< n->data << " ";
        n = n->next;
    }
    cout<< endl;
}


// } Driver Code Ends
/* node for linked list:

struct Node {
    int data;
    struct Node* next;
    Node(int x) {
        data = x;
        next = NULL;
    }
};

*/

class Solution
{
    public:
    Node* reverse(Node* head){
        Node* prev=NULL;
        Node* curr=head;
        while(curr!=NULL){
            Node* forward=curr->next;
            curr->next=prev;
            prev=curr;
            curr=forward;
        }
        return prev;
    }
    Node* add(Node* first,Node* second){
        Node* ans=new Node(0);
        Node* temp=ans;
        int carry=0,sum=0;
        while(first!=NULL && second!=NULL){
            sum = carry+ first->data + second->data;
            carry=sum/10;
            int digit=sum%10;
            ans->next=new Node(digit);
            ans=ans->next;
            first=first->next;
            second=second->next;
        }
        while(first!=NULL){
            sum = carry+ first->data;
            carry=sum/10;
            int digit=sum%10;
            ans->next=new Node(digit);
            ans=ans->next;
            first=first->next;
        }
        while(second!=NULL){
            sum = carry+ second->data;
            carry=sum/10;
            int digit=sum%10;
            ans->next=new Node(digit);
            ans=ans->next;
            second=second->next;
        }
        if(carry!=0){
            ans->next=new Node(carry);
            ans=ans->next;
        }
        return temp->next;
    }
    //Function to add two numbers represented by linked list.
    struct Node* addTwoLists(struct Node* first, struct Node* second)
    {
        // first reverse both the linked list
        first=reverse(first);
        second=reverse(second);
        
        // this is the function for adding the digits
        Node* ans=add(first,second);
        // then reverse the ans linked list and return the head of the ans linked list
        ans=reverse(ans);
        return ans;
        
    }
};


//{ Driver Code Starts.

int main()
{
    int t;
    cin>>t;
    while(t--)
    {
        int n, m;
        
        cin>>n;
        Node* first = buildList(n);
        
        cin>>m;
        Node* second = buildList(m);
        Solution ob;
        Node* res = ob.addTwoLists(first,second);
        printList(res);
    }
    return 0;
}

// } Driver Code Ends

File: Medium/Add two numbers represented by linked lists/add-two-numbers-represented-by-linked-lists.py
#User function Template for python3

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class Solution:
    #Function to add two numbers represented by linked list.
    def addTwoLists(self, num1, num2):
        
        curr1 , curr2 = num1 , num2 
        
        res1 , res2 = "" , ""
        while curr1:
            res1+=str(curr1.data)
            curr1 = curr1.next 
        while curr2:
            res2+=str(curr2.data)
            curr2 = curr2.next 
        
        
        res1 , res2  = res1 , res2
        
        res1 , res2 = int(res1) , int(res2)
        
        res = res1 + res2
        
        res = str(res)[::-1]
        
        
        head = None
        
        for i in res:
            new_node = Node(i)
            new_node.next = head
            head = new_node
            
        return head
        


#{ 
 # Driver Code Starts
#Initial Template for Python 3

# Node Class
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

# Linked List Class
class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    # creates a new node with given value and appends it at the end of the linked list
    def insert(self, val):
        if self.head is None:
            self.head = Node(val)
            self.tail = self.head
        else:
            self.tail.next = Node(val)
            self.tail = self.tail.next

# prints the elements of linked list starting with head
def printList(n):
    while n:
        print(n.data,end=' ')
        n = n.next
    print()

if __name__ == '__main__':
    for _ in range(int(input())):
        
        n = int(input())
        arr1 = ( int(x) for x in input().split() )
        num1 = LinkedList()
        for i in arr1:
            num1.insert(i)
        
        m = int(input())
        arr2 = ( int(x) for x in input().split() )
        num2 = LinkedList()
        for i in arr2:
            num2.insert(i)
        
        res = Solution().addTwoLists(num1.head, num2.head)
        printList(res)
# } Driver Code Ends